sdk/generic-sdk-core/generic-core-http-okhttp/src/main/java/com/generic/core/http/okhttp/OkHttpHttpClient.java

```diff
@@ -138,40 +141,74 @@ private static long getRequestContentLength(BinaryData content, HttpHeaders head
         return contentLength;
     }
 
-    private Response<?> toResponse(HttpRequest request, okhttp3.Response response, boolean eagerlyReadResponse,
-                                   boolean ignoreResponseBody, boolean eagerlyConvertHeaders) throws IOException {
+    private Response<?> toResponse(HttpRequest request, okhttp3.Response response,
+                                   ResponseBodyMode responseBodyMode, boolean eagerlyConvertHeaders) throws IOException {
         okhttp3.Headers responseHeaders = response.headers();
 
         if (isTextEventStream(responseHeaders) && response.body() != null) {
             ServerSentEventListener listener = request.getServerSentEventListener();
+
             if (listener != null) {
-                processTextEventStream(request,
-                    httpRequest -> this.send(httpRequest), response.body().byteStream(), listener, LOGGER);
+                processTextEventStream(request, this::send, response.body().byteStream(), listener, LOGGER);
             } else {
                 throw LOGGER.logThrowableAsError(new RuntimeException(ServerSentEventUtil.NO_LISTENER_ERROR_MESSAGE));
             }
-            return new OkHttpResponse(response, request, eagerlyConvertHeaders, EMPTY_BODY);
+
+            return new OkHttpResponse(response, request, eagerlyConvertHeaders, BinaryData.fromBytes(EMPTY_BODY));
         }
-        return processResponse(request, response, eagerlyReadResponse, ignoreResponseBody, eagerlyConvertHeaders);
+
+        return processResponse(request, response, responseBodyMode, eagerlyConvertHeaders);
     }
 
-    private Response<?> processResponse(HttpRequest request, okhttp3.Response response, boolean eagerlyReadResponse,
-                                        boolean ignoreResponseBody, boolean eagerlyConvertHeaders) throws IOException {
-        if (eagerlyReadResponse || ignoreResponseBody) {
-            try (ResponseBody body = response.body()) {
-                byte[] bytes = (body != null) ? body.bytes() : EMPTY_BODY;
-                return new OkHttpResponse(response, request, eagerlyConvertHeaders, bytes);
+    private Response<?> processResponse(HttpRequest request, okhttp3.Response response,
+                                        ResponseBodyMode responseBodyMode, boolean eagerlyConvertHeaders) throws IOException {
+        if (responseBodyMode == null) {
+            String contentType = response.headers().get(CONTENT_TYPE.getCaseInsensitiveName());
+
+            if (APPLICATION_OCTET_STREAM.equalsIgnoreCase(contentType)) {
+                responseBodyMode = STREAM;
+            } else {
+                responseBodyMode = BUFFER;
             }
-        } else {
-            return new OkHttpResponse(response, request, eagerlyConvertHeaders, null);
+
+            request.getMetadata().setResponseBodyMode(responseBodyMode);
```

Comment:
Should we be updating the request or is this something that should be set in the response? Setting this in the response may make more sense as it's used for it and could be made immutable.

Reply:
Since we don't have anything like `ResponseOptions`, we set this value in `RequestOptions` and later on `HttpRequestMetadata` to get it plumbed through to `SwaggerMethodParser` and other classes that deal with getting and handling the `Response`.

We can certainly set the mode on `Response` as well, but I don't see much value in users knowing how the response body was handled after they get their hands on it. The way we've been doing things so far makes me think we want to keep the APIs in `Response` to a minimum.

Reply:
I'm confused - why are we modifying the request at this point? Maybe this is just because at this point we haven't done the work to remove HttpRequestMetadata and plumb RequestOptions in properly, but regardless, we shouldn't modify state like this. This should be marked as a FIXME that gets fixed when HttpRequestMetadata is removed.